File: Forms/ChooseStateClassForm.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Imports SyncroSim.Core
Imports System.Windows.Forms

''' <summary>
''' Choose State Class form
''' </summary>
''' <remarks></remarks>
Class ChooseStateClassForm

    Private m_Diagram As TransitionDiagram
    Private m_TerminologyDataSheet As DataSheet
    Private m_StateLabelXDataSheet As DataSheet
    Private m_StateLabelYDataSheet As DataSheet
    Private m_ChosenStateLabelX As BaseValueDisplayListItem
    Private m_ChosenStateLabelY As BaseValueDisplayListItem
    Private m_EditMode As Boolean

    ''' <summary>
    ''' Gets the chosen state label x combo item
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property ChosenStateLabelX As BaseValueDisplayListItem
        Get
            Return Me.m_ChosenStateLabelX
        End Get
    End Property

    ''' <summary>
    ''' Gets the chosen state label y combo item
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property ChosenStateLabelY As BaseValueDisplayListItem
        Get
            Return Me.m_ChosenStateLabelY
        End Get
    End Property

    ''' <summary>
    ''' Initializes this form
    ''' </summary>
    ''' <param name="Diagram"></param>
    ''' <param name="dataFeed"></param>
    ''' <param name="editMode"></param>
    ''' <remarks></remarks>
    Public Function Initialize(
        ByVal diagram As TransitionDiagram,
        ByVal dataFeed As DataFeed,
        ByVal editMode As Boolean) As Boolean

        Me.m_Diagram = diagram

        Me.m_TerminologyDataSheet = dataFeed.Project.GetDataSheet(DATASHEET_TERMINOLOGY_NAME)
        Me.m_StateLabelXDataSheet = dataFeed.Project.GetDataSheet(DATASHEET_STATE_LABEL_X_NAME)
        Me.m_StateLabelYDataSheet = dataFeed.Project.GetDataSheet(DATASHEET_STATE_LABEL_Y_NAME)

        Me.FillComboBoxes()

        If (Me.ComboBoxStateLabelX.Items.Count = 0) Then

            FormsUtilities.ErrorMessageBox(ERROR_DIAGRAM_NO_STATE_LABEL_X_VALUES)
            Return False

        ElseIf (Me.ComboBoxStateLabelY.Items.Count = 0) Then

            FormsUtilities.ErrorMessageBox(ERROR_DIAGRAM_NO_STATE_LABEL_Y_VALUES)
            Return False

        End If

        Me.ResetSlxAndSlyLabels()
        Me.m_EditMode = editMode

        If (Me.m_EditMode) Then

            Me.Text = "Edit State Class"

            Dim EditShape As StateClassShape = CType(diagram.SelectedShapes.First, StateClassShape)
            Me.SelectComboValues(EditShape.StateLabelXId, EditShape.StateLabelYId)

        Else
            Me.Text = "Add State Class"
        End If

        Return True

    End Function

    ''' <summary>
    ''' Fills the combo boxes
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub FillComboBoxes()

        Dim dvct As New DataView(Me.m_StateLabelXDataSheet.GetData(), Nothing,
            Me.m_StateLabelXDataSheet.DisplayMember, DataViewRowState.CurrentRows)

        For Each drv As DataRowView In dvct

            Dim Id As Integer = CInt(drv.Row(Me.m_StateLabelXDataSheet.ValueMember))
            Dim Name As String = CStr(drv.Row(Me.m_StateLabelXDataSheet.DisplayMember))

            Me.ComboBoxStateLabelX.Items.Add(New BaseValueDisplayListItem(Id, Name))

        Next

        Dim dvss As New DataView(Me.m_StateLabelYDataSheet.GetData(), Nothing,
            Me.m_StateLabelYDataSheet.DisplayMember, DataViewRowState.CurrentRows)

        For Each drv As DataRowView In dvss

            Dim Id As Integer = CInt(drv.Row(Me.m_StateLabelYDataSheet.ValueMember))
            Dim Name As String = CStr(drv.Row(Me.m_StateLabelYDataSheet.DisplayMember))

            Me.ComboBoxStateLabelY.Items.Add(New BaseValueDisplayListItem(Id, Name))

        Next

        If (Me.ComboBoxStateLabelX.Items.Count > 0) Then
            Me.ComboBoxStateLabelX.SelectedIndex = 0
        End If

        If (Me.ComboBoxStateLabelY.Items.Count > 0) Then
            Me.ComboBoxStateLabelY.SelectedIndex = 0
        End If

    End Sub

    ''' <summary>
    ''' Selects the specified combo values
    ''' </summary>
    ''' <param name="stateLabelXId"></param>
    ''' <param name="stateLabelYId"></param>
    ''' <remarks></remarks>
    Private Sub SelectComboValues(ByVal stateLabelXId As Integer, ByVal stateLabelYId As Integer)

        Debug.Assert(Me.m_EditMode)

        For Each item As BaseValueDisplayListItem In Me.ComboBoxStateLabelX.Items

            If (item.Value = stateLabelXId) Then

                Me.ComboBoxStateLabelX.SelectedItem = item
                Exit For

            End If

        Next

        For Each item As BaseValueDisplayListItem In Me.ComboBoxStateLabelY.Items

            If (item.Value = stateLabelYId) Then

                Me.ComboBoxStateLabelY.SelectedItem = item
                Exit For

            End If

        Next

    End Sub

    ''' <summary>
    ''' Resets the state label x and y form labels
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub ResetSlxAndSlyLabels()

        Dim slxlabel As String = Nothing
        Dim slylabel As String = Nothing

        GetStateLabelTerminology(Me.m_TerminologyDataSheet, slxlabel, slylabel)

        Me.LabelStateLabelX.Text = slxlabel & ":"
        Me.LabelStateLabelY.Text = slylabel & ":"

    End Sub

    ''' <summary>
    ''' Determines if the specified state class exists in the diagram
    ''' </summary>
    ''' <param name="slxid"></param>
    ''' <param name="slyid"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Private Function StateClassInDiagram(ByVal slxid As Integer, ByVal slyid As Integer) As Boolean

        For Each s As StateClassShape In Me.m_Diagram.Shapes

            If (Me.m_Diagram.StratumId.HasValue) Then

                If (Not s.StratumIdSource.HasValue) Then
                    Continue For
                End If

            End If

            If (s.StateLabelXId = slxid And s.StateLabelYId = slyid) Then
                Return True
            End If

        Next

        Return False

    End Function

    ''' <summary>
    ''' Handles the OK button Clicked event
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub ButtonOK_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ButtonOK.Click

        Me.m_ChosenStateLabelX = CType(Me.ComboBoxStateLabelX.Items(
                Me.ComboBoxStateLabelX.SelectedIndex), BaseValueDisplayListItem)

        Me.m_ChosenStateLabelY = CType(Me.ComboBoxStateLabelY.Items(
                Me.ComboBoxStateLabelY.SelectedIndex), BaseValueDisplayListItem)

        If (Not Me.m_EditMode) Then

            If (Me.StateClassInDiagram(Me.m_ChosenStateLabelX.Value, Me.m_ChosenStateLabelY.Value)) Then

                FormsUtilities.ErrorMessageBox(ERROR_DIAGRAM_STATE_CLASS_EXISTS)
                Return

            End If

        End If

        Me.DialogResult = DialogResult.OK
        Me.Close()

    End Sub

    ''' <summary>
    ''' Handles the Cancel button Clicked event
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub ButtonCancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ButtonCancel.Click

        Me.DialogResult = DialogResult.Cancel
        Me.Close()

    End Sub

End Class

File: Forms/ChooseStateClassForm.cs
// ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System;
using System.Data;
using System.Linq;
using SyncroSim.Core;
using System.Diagnostics;

namespace SyncroSim.STSim
{
    internal partial class ChooseStateClassForm
    {
        public ChooseStateClassForm()
        {
            InitializeComponent();
        }

        private TransitionDiagram m_Diagram;
        private DataSheet m_TerminologyDataSheet;
        private DataSheet m_StateLabelXDataSheet;
        private DataSheet m_StateLabelYDataSheet;
        private BaseValueDisplayListItem m_ChosenStateLabelX;
        private BaseValueDisplayListItem m_ChosenStateLabelY;
        private bool m_EditMode;

        public BaseValueDisplayListItem ChosenStateLabelX
        {
            get
            {
                return this.m_ChosenStateLabelX;
            }
        }

        public BaseValueDisplayListItem ChosenStateLabelY
        {
            get
            {
                return this.m_ChosenStateLabelY;
            }
        }

        public bool Initialize(TransitionDiagram diagram, DataFeed dataFeed, bool editMode)
        {
            this.m_Diagram = diagram;

            this.m_TerminologyDataSheet = dataFeed.Project.GetDataSheet(Strings.DATASHEET_TERMINOLOGY_NAME);
            this.m_StateLabelXDataSheet = dataFeed.Project.GetDataSheet(Strings.DATASHEET_STATE_LABEL_X_NAME);
            this.m_StateLabelYDataSheet = dataFeed.Project.GetDataSheet(Strings.DATASHEET_STATE_LABEL_Y_NAME);

            this.FillComboBoxes();

            if (this.ComboBoxStateLabelX.Items.Count == 0)
            {
                FormsUtilities.ErrorMessageBox(MessageStrings.ERROR_DIAGRAM_NO_STATE_LABEL_X_VALUES);
                return false;
            }
            else if (this.ComboBoxStateLabelY.Items.Count == 0)
            {
                FormsUtilities.ErrorMessageBox(MessageStrings.ERROR_DIAGRAM_NO_STATE_LABEL_Y_VALUES);
                return false;
            }

            this.ResetSlxAndSlyLabels();
            this.m_EditMode = editMode;

            if (this.m_EditMode)
            {
                this.Text = "Edit State Class";

                StateClassShape EditShape = (StateClassShape)diagram.SelectedShapes.First();
                this.SelectComboValues(EditShape.StateLabelXId, EditShape.StateLabelYId);
            }
            else
            {
                this.Text = "Add State Class";
            }

            return true;
        }

        private void FillComboBoxes()
        {
            DataView dvct = new DataView(
                this.m_StateLabelXDataSheet.GetData(), 
                null, 
                this.m_StateLabelXDataSheet.DisplayMember, 
                DataViewRowState.CurrentRows);

            foreach (DataRowView drv in dvct)
            {
                int Id = Convert.ToInt32(drv.Row[this.m_StateLabelXDataSheet.ValueMember]);
                string Name = Convert.ToString(drv.Row[this.m_StateLabelXDataSheet.DisplayMember]);

                this.ComboBoxStateLabelX.Items.Add(new BaseValueDisplayListItem(Id, Name));
            }

            DataView dvss = new DataView(
                this.m_StateLabelYDataSheet.GetData(), 
                null, 
                this.m_StateLabelYDataSheet.DisplayMember, 
                DataViewRowState.CurrentRows);

            foreach (DataRowView drv in dvss)
            {
                int Id = Convert.ToInt32(drv.Row[this.m_StateLabelYDataSheet.ValueMember]);
                string Name = Convert.ToString(drv.Row[this.m_StateLabelYDataSheet.DisplayMember]);

                this.ComboBoxStateLabelY.Items.Add(new BaseValueDisplayListItem(Id, Name));
            }

            if (this.ComboBoxStateLabelX.Items.Count > 0)
            {
                this.ComboBoxStateLabelX.SelectedIndex = 0;
            }

            if (this.ComboBoxStateLabelY.Items.Count > 0)
            {
                this.ComboBoxStateLabelY.SelectedIndex = 0;
            }
        }

        private void SelectComboValues(int stateLabelXId, int stateLabelYId)
        {
            Debug.Assert(this.m_EditMode);

            foreach (BaseValueDisplayListItem item in this.ComboBoxStateLabelX.Items)
            {
                if (item.Value == stateLabelXId)
                {
                    this.ComboBoxStateLabelX.SelectedItem = item;
                    break;
                }
            }

            foreach (BaseValueDisplayListItem item in this.ComboBoxStateLabelY.Items)
            {
                if (item.Value == stateLabelYId)
                {
                    this.ComboBoxStateLabelY.SelectedItem = item;
                    break;
                }
            }
        }

        private void ResetSlxAndSlyLabels()
        {
            string slxlabel = null;
            string slylabel = null;

            TerminologyUtilities.GetStateLabelTerminology(this.m_TerminologyDataSheet, ref slxlabel, ref slylabel);

            this.LabelStateLabelX.Text = slxlabel + ":";
            this.LabelStateLabelY.Text = slylabel + ":";
        }

        private bool StateClassInDiagram(int slxid, int slyid)
        {
            foreach (StateClassShape s in this.m_Diagram.Shapes)
            {
                if (this.m_Diagram.StratumId.HasValue)
                {
                    if (!s.StratumIdSource.HasValue)
                    {
                        continue;
                    }
                }

                if (s.StateLabelXId == slxid && s.StateLabelYId == slyid)
                {
                    return true;
                }
            }

            return false;
        }

        private void ButtonOK_Click(object sender, System.EventArgs e)
        {
            this.m_ChosenStateLabelX = (BaseValueDisplayListItem)this.ComboBoxStateLabelX.Items[this.ComboBoxStateLabelX.SelectedIndex];
            this.m_ChosenStateLabelY = (BaseValueDisplayListItem)this.ComboBoxStateLabelY.Items[this.ComboBoxStateLabelY.SelectedIndex];

            if (!this.m_EditMode)
            {
                if (this.StateClassInDiagram(this.m_ChosenStateLabelX.Value, this.m_ChosenStateLabelY.Value))
                {
                    FormsUtilities.ErrorMessageBox(MessageStrings.ERROR_DIAGRAM_STATE_CLASS_EXISTS);
                    return;
                }
            }

            this.DialogResult = System.Windows.Forms.DialogResult.OK;
            this.Close();
        }

        private void ButtonCancel_Click(object sender, System.EventArgs e)
        {
            this.DialogResult = System.Windows.Forms.DialogResult.Cancel;
            this.Close();
        }
    }
}
